modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/ControlAcceleratorSupport.java

```diff
@@ -132,23 +133,32 @@ private static ChangeListener<Scene> getSceneChangeListener(Object anchor, Obser
         return sceneChangeListener;
     }
 
+    /* It's okay to have the value Weak, because we only remember it to remove the listener later on */
+    private static Map<ListChangeListener<MenuItem>, WeakReference<ListChangeListener<MenuItem>>>
+            menuListChangeListenerMap = new WeakHashMap<>();
+
     private static void doAcceleratorInstall(final ObservableList<MenuItem> items, final Scene scene) {
         // we're given an observable list of menu items, which we will add an observer to
         // so that when menu items are added or removed we can properly handle
         // the addition or removal of accelerators into the scene.
-        items.addListener((ListChangeListener<MenuItem>) c -> {
-            while (c.next()) {
-                if (c.wasRemoved()) {
-                    // remove accelerators from the scene
-                    removeAcceleratorsFromScene(c.getRemoved(), scene);
-                }
+        ListChangeListener<MenuItem> listChangeListener = new IdentityWrapperListChangeListener(items) {
+            @Override
+            public void onChanged(Change<? extends MenuItem> c) {
+                while (c.next()) {
+                    if (c.wasRemoved()) {
+                        // remove accelerators from the scene
+                        removeAcceleratorsFromScene(c.getRemoved(), scene);
+                    }
 
-                if (c.wasAdded()) {
-                    ControlAcceleratorSupport.doAcceleratorInstall(c.getAddedSubList(), scene);
+                    if (c.wasAdded()) {
+                        ControlAcceleratorSupport.doAcceleratorInstall(c.getAddedSubList(), scene);
+                    }
                 }
             }
-        });
+        };
 
+        menuListChangeListenerMap.put(listChangeListener, new WeakReference<>(listChangeListener));
```

Comment:
is it possible this method will be called twice with the same (items, scene) arguments?  i.e. is it possible to clobber an already registered listChangeListener?

Reply:
Maybe. I could check if one's there and remove it? If it happened, there would be more change listeners on the list, but only 1 in the HashMap, and when removed, only one of the copies would be removed.

I'm not sure it can happen because of the checks in the methods that call this method. It hasn't come up in my tests, but I also don't think there's harm in removing one if there is before.